[WFLY-19954] Add the vertx extension/subsystem from wildfly-vertx-feature-pack to WildFly Preview Feature Pack

gaol opened the pull request:
Issue: https://issues.redhat.com/browse/WFLY-19954

Resolve: https://github.com/wildfly/wildfly-proposals/issues/662

gaol committed:
[WFLY-19954] Add the wildfly-vertx-feature-pack to WildFly Preview

gaol commented:
@kabir @jasondlee Would you please review ? thanks.

bstansberry commented:
Hi @gaol -- FYI re https://github.com/wildfly/wildfly/pull/18466. If that solves the WFLY-20035 problem the vertx extension will likely need similar handling.

kabir commented on testsuite/preview/basic/src/test/java/org/wildfly/test/preview/vertx/VertxSubsystemSmokeTestCase.java:
```diff
@@ -0,0 +1,129 @@
+    public void testFail() throws Exception {
+        try {
+            HttpRequest.get(url.toExternalForm() + "rest/echo/fail/Hello", 4, TimeUnit.SECONDS);
+            Assert.fail("Not here");
```
s/Not here/Should have failed

gaol replied:
> My guess is that `vertxInstance.select(Identifier.Literal.of("vertx-not-exist")).get()` throws some kind of exception, and thus the GET in `VertxSubsystemSmokeTestCase.testFail()` fails ?

Yes, it is, the CDI exception(`org.jboss.weld.exceptions.UnsatisfiedResolutionException`
) gets propagated to client via `return Response.status(500).entity(e).build()` and should contain 
 a specific message which is checked in the test:

```java
Assert.assertTrue(e.getMessage().contains("WELD-001334: Unsatisfied dependencies for type Vertx"));
```

kabir reviewed (commented):
Hi @gaol 

I'm not super-familiar with CDI, so I just wanted to double check what happens in the `!exist` case in this EchoService method:

```
private Vertx getVertx(boolean exist) {
        if (exist) {
            return vertxInstance.select(Identifier.Literal.of("vertx")).get();
        }
        return vertxInstance.select(Identifier.Literal.of("vertx-not-exist")).get();
    }
```

My guess is that `vertxInstance.select(Identifier.Literal.of("vertx-not-exist")).get()` throws some kind of exception, and thus the GET in `VertxSubsystemSmokeTestCase.testFail()` fails ?

It looks like that reading the code, just wanted to double check :-)

gaol commented:
> Hi @gaol -- FYI re #18466. If that solves the WFLY-20035 problem the vertx extension will likely need similar handling.

yes, I created https://github.com/wildfly-extras/wildfly-vertx-feature-pack/issues/46 to track the changes in the vertx extension and will update this one later after another release.

kabir commented on testsuite/preview/basic/src/test/java/org/wildfly/test/preview/vertx/VertxSubsystemSmokeTestCase.java:
```diff
@@ -0,0 +1,129 @@
+            HttpRequest.get(url.toExternalForm() + "rest/echo/fail/Hello", 4, TimeUnit.SECONDS);
+            Assert.fail("Should have failed");
+        } catch (IOException e) {
+            Assert.assertTrue(e.getMessage().contains("WELD-001334: Unsatisfied dependencies for type Vertx"));
```
Actually, here I guess this message could be internationalised, so it is probably better to just check for "WELD-001334"

gaol replied:
Ah, yes. thank you. will update it  :)

gaol commented:
Thank you @bstansberry @kabir for the comments,  I updated this PR with:
* Upgrade the wildfly-vertx-feature-pack to 1.0.2.Final with https://github.com/wildfly-extras/wildfly-vertx-feature-pack/issues/46 addressed
* Update the test code to check the exception id only.

bstansberry commented:
@jasondlee @fl4via Please review

jasondlee reviewed (approved):
lgtm

gaol committed:
Adds io.smallrye.common.annotation dependency to io.smallrye.opentelemetry module

gaol committed:
Set vertx.cdi.identifier for opentelemetry and reactive-messaging subsystems

gaol commented:
As this PR has conflicts and smallrye-opentelemetry has been upgraded to 2.9.0, I updated this PR with additional commits:

* Fixed code conflicts
* Added `io.smallrye.common.annotation` to the dependency of `io.smallrye.opentelemetry` module
* Put `otel.exporter.vertx.cdi.identifier` with value of `vertx` to OpenTelemetryConfig
* Put `mp.messaging.connector.vertx.cdi.identifier` with value of `vertx` to ReactiveMessagingConfigSource

The latter 2 changes make the 2 subsystems try to lookup the Vertx instance from the CDI identifier, if there is no one exposed from vertx subsystem, it falls back to what they currently do(call Vertx.vertx())
